internal/db/schema/migrations/postgres/12/03_vault_credential.up.sql

```diff
@@ -55,26 +57,90 @@ begin;
   create trigger delete_credential_store_subtype after delete on credential_vault_store
     for each row execute procedure delete_credential_store_subtype();
 
+  -- before_soft_delete_credential_vault_store is a before update trigger for
+  -- the credential_vault_store table that makes the delete_time column a
+  -- set once column. Once the delete_time column is set to a value other than
+  -- null, it cannot be changed. If the current delete_time of a row is not null
+  -- and an update contains a value for delete_time different from the current
+  -- value, this trigger will raise an error with error code 23602 which is a
+  -- class 23 integrity constraint violation: set_once_violation.
+  create function before_soft_delete_credential_vault_store()
+    returns trigger
+  as $$
+  begin
+    if new.delete_time is distinct from old.delete_time then
+      if old.delete_time is not null then
+        raise exception 'set_once_violation: %.%', tg_table_name, 'delete_time' using
+          errcode = '23602',
+          schema = tg_table_schema,
+          table = tg_table_name,
+          column = 'delete_time';
+      end if;
+    end if;
+    return new;
+  end;
+  $$ language plpgsql;
+
+  create trigger before_soft_delete_credential_vault_store before update on credential_vault_store
+    for each row execute procedure before_soft_delete_credential_vault_store();
+
+  -- after_soft_delete_credential_vault_store is an after update trigger for the
+  -- credential_vault_store table that performs cleanup actions when a
+  -- credential store is soft deleted. A credential store is considered "soft
+  -- deleted" if the delete_time for the row is updated from null to not null.
+  --
+  -- When a credential store is soft deleted, this trigger:
+  --  * marks any active Vault tokens owned by the credential store for revocation
+  --  * deletes any credential library owned by the credential store
+  create function after_soft_delete_credential_vault_store()
+    returns trigger
+  as $$
+  begin
+    if new.delete_time is distinct from old.delete_time then
+      if old.delete_time is null then
+
+        -- mark current and maintaining tokens as revoke
+        update credential_vault_token
+           set status   = 'revoke'
+         where store_id = new.public_id
+           and status in ('current', 'maintaining');
+
+        -- delete the store's libraries
+        delete
```

Comment:
This will by-pass the cred lib delete oplog

Reply:
I think that should be fine since the same trigger will execute when the oplog is replayed.

Reply:
Aha yes great point!